B-18032: Allow an Update for an approved service item

TevinAdams opened the pull request:
## [https://www13.v1host.com/USTRANSCOM38/Default.aspx?menu=StoryBoardPage](B-18032)

## Summary

This change allows fields sitDepartureDate, sitRequestedDelivery, sitCustomerContacted, and sitDestinationFinalAddress to be updated when the a DOASIT, DOFSIT, DOPSIT, DDDSIT serviceItem is in an APPROVED status.

## Testing
To test try updating one of the fields listed above on a  DOASIT, DOFSIT, DDDSIT or DOPSIT serviceItem when in approved status.

danieljordan-caci commented:
So... When using the API and the payload is being converted to a model in `payload_to_model.go`, it seems like the service code is not being passed into that so the checks that are happening on line 595 aren't happening. Specifically it's checking if the service item is a destination SIT item and then pulls the additional values needed, which we need:
`sitCustomerContacted` & `sitRequestedDelivery`

In my PR here: https://github.com/transcom/mymove/pull/11638/files#diff-54081105a3cca8353f6085cc2d26c4c98b51c1e0622cb0b48fdc26a93335a304

I added in populating those values prior to those checks because the gov wanted origin SIT items to also have those values updated.

So it seems that the service code is not being passed into the `payload_to_model.go` function when it is of model `UpdateMTOServiceItemSIT`, which is not populating those two values.

If my PR merges then this issue will go away because I added those prior to the service code check, but the problem still persists... How are we passing in the service code prior to converting the payload to a model?

danieljordan-caci commented on pkg/services/mto_service_item/mto_service_item_validators.go:
```diff
@@ -234,10 +234,61 @@ func (v *updateMTOServiceItemData) checkOldServiceItemStatus(_ appcontext.AppCon
+				invalidFieldChange = true
+			}
+
+			if invalidFieldChange {
```
Also... are you opposed to doing these if checks like this?

```
	invalidFieldChange := false

			// Fields that are not allowed to change when status is approved
			if serviceItemData.updatedServiceItem.ReService.Code.String() != "" ||
				serviceItemData.updatedServiceItem.SITEntryDate != nil ||
				serviceItemData.updatedServiceItem.Reason != nil ||
				serviceItemData.updatedServiceItem.SITPostalCode != nil ||
				serviceItemData.updatedServiceItem.RequestedApprovalsRequestedStatus != nil {
				invalidFieldChange = true
			}

			if invalidFieldChange {
				return apperror.NewConflictError(serviceItemData.oldServiceItem.ID,
					"- one or more fields is not allowed to be updated when the SIT service item has an approved status.")
			}

			// Fields allowed to be changed when status is approved
			if serviceItemData.updatedServiceItem.SITDepartureDate != nil ||
				serviceItemData.updatedServiceItem.SITDestinationFinalAddress != nil ||
				serviceItemData.updatedServiceItem.SITRequestedDelivery != nil ||
				serviceItemData.updatedServiceItem.SITCustomerContacted != nil {
				return nil
			}

			return apperror.NewConflictError(serviceItemData.oldServiceItem.ID,
				"- unkown field or fields attempting to be updated.")
```

TevinAdams replied:
I am not. I spent some time making a more robust way but it would make current implementation useless. I basically was going to take another approach that would track changes and a few other things. It was taking more time and would cause more time so I just kept it simple. I didn't want to break current functionality and change current implementation so if the status is approved it goes through a separate status check than if rejected. That was the most simplest and quickest solution for the sake of time.

danieljordan-caci replied:
Whatever you wanna do! I think performance wise, they are both the same - I just prefer that way cause it looks more concise and easier to add checks if we need to in the future (simple one line change).

I'll leave it up to you though! It def still works with your implementation.

danieljordan-caci replied:
You're good! It really doesn't matter, was just showing you what I prefer. They both do the same job successfully - it's your PR so I'll let you roll with whatever you prefer!

TevinAdams commented:
> > So... When using the API and the payload is being converted to a model in `payload_to_model.go`, it seems like the service code is not being passed into that so the checks that are happening on line 595 aren't happening. Specifically it's checking if the service item is a destination SIT item and then pulls the additional values needed, which we need: `sitCustomerContacted` & `sitRequestedDelivery`
> > In my PR here: https://github.com/transcom/mymove/pull/11638/files#diff-54081105a3cca8353f6085cc2d26c4c98b51c1e0622cb0b48fdc26a93335a304
> > I added in populating those values prior to those checks because the gov wanted origin SIT items to also have those values updated.
> > So it seems that the service code is not being passed into the `payload_to_model.go` function when it is of model `UpdateMTOServiceItemSIT`, which is not populating those two values.
> > If my PR merges then this issue will go away because I added those prior to the service code check, but the problem still persists... How are we passing in the service code prior to converting the payload to a model?
> 
> And just to summarize this... This means that we aren't able to update `sitCustomerContacted` & `sitRequestedDelivery` until my PR merges.

We can probably meet about this. Im trying to understand what your talking about and working on.

traskowskycaci reviewed (commented):
I am able to update the SIT departure date, SIT requested delivery, and SIT customer contacted successfully, but haven't been able to update the SIT destination final address. I have tried using a string value and also an address id in the body and get a 500 response of:
`{
  "title": "Internal Server Error",
  "instance": "8d24fb7d-b5b2-490d-a2db-98a58840ae36",
  "detail": "json: cannot unmarshal string into Go struct field .sitDestinationFinalAddress of type primemessages.Address"
}`

traskowskycaci commented:
> I am able to update the SIT departure date, SIT requested delivery, and SIT customer contacted successfully, but haven't been able to update the SIT destination final address. I have tried using a string value and also an address id in the body and get a 500 response of:

Should have clarified that I was using the swagger prime API. The issue with updating the destination final address may not be an issue based on the latest developments?

danieljordan-caci commented:
> > I am able to update the SIT departure date, SIT requested delivery, and SIT customer contacted successfully, but haven't been able to update the SIT destination final address. I have tried using a string value and also an address id in the body and get a 500 response of:
> 
> Should have clarified that I was using the swagger prime API. The issue with updating the destination final address may not be an issue based on the latest developments?

Correct - it's how the updater is handling the address. Only updates that value if the address does not exist yet. Doesn't update it.

TevinAdams commented:
> > I am able to update the SIT departure date, SIT requested delivery, and SIT customer contacted successfully, but haven't been able to update the SIT destination final address. I have tried using a string value and also an address id in the body and get a 500 response of:
> 
> Should have clarified that I was using the swagger prime API. The issue with updating the destination final address may not be an issue based on the latest developments?

We will not be allowing SIT destination final address updates per conversations with DAN today.

traskowskycaci reviewed (approved):
Looks good for all of those service item codes and for all three fields getting updated in the database! Tested all with Swagger.
